test-cypress/cypress/commands/api/visits.ts

```diff
@@ -5,39 +5,117 @@ import { v1ApiPath, getCreds } from "../server";
-  (user: string, camera: string, noVists: number) => {
-    logTestDescription(`Check number of visits is ${noVists}`, {
-      user: user,
-      camera: camera,
-      expectedVisits: noVists
+  "checkVisitTags",
+  (user: string, camera: string, expectedTags: string[]) => {
+    const expectedVisits = expectedTags.map((tag) => {
+      return { tag };
     });
 
-    const where: VisitsWhere = {
-      duration: { $gte: "0" },
-      type: "thermalRaw"
-    };
+    logTestDescription(
+      `Check visit tags match ${JSON.stringify(expectedTags)}`,
+      {
+        user,
+        camera,
+        expectedVisits
+      }
+    );
 
-    if (camera) {
-      where.DeviceId = getCreds(camera).id;
-    }
+    checkVisitsMatch(user, camera, expectedVisits);
+  }
+);
 
-    const params = {
-      where: JSON.stringify(where),
-      limit: 100
-    };
-
-    const visitsNum = noVists;
-    cy.request({
-      method: "GET",
-      url: v1ApiPath("recordings/visits", params),
-      headers: getCreds(user).headers
-    }).then((response) => {
-      expect(response.body.numVisits).to.eq(visitsNum);
+Cypress.Commands.add(
+  "checkVisits",
+  (user: string, camera: string, expectedVisits: ComparableVisit[]) => {
+    logTestDescription(`Check visits match ${JSON.stringify(expectedVisits)}`, {
+      user,
+      camera,
+      expectedVisits
     });
+
+    checkVisitsMatch(user, camera, expectedVisits);
   }
 );
 
+function checkVisitsMatch(
+  user: string,
+  camera: string,
+  expectedVisits: ComparableVisit[]
+) {
+  const where: VisitsWhere = {
+    duration: { $gte: "0" },
+    type: "thermalRaw"
+  };
+
+  if (camera) {
+    where.DeviceId = getCreds(camera).id;
+  }
+
+  const params = {
+    where: JSON.stringify(where),
+    limit: 100
+  };
+
+  cy.request({
+    method: "GET",
+    url: v1ApiPath("recordings/visits", params),
+    headers: getCreds(user).headers
+  }).then((response) => {
+    checkResponseMatches(response, expectedVisits);
+  });
+}
+
+function checkResponseMatches(
+  response: any,
+  expectedVisits: ComparableVisit[]
+) {
+  const responseVisits = response.body.visits;
+
+  expect(
+    responseVisits.length,
+    `Number of visits to be ${responseVisits.length}`
+  ).to.eq(expectedVisits.length);
+  const increasingDateResponseVisits = responseVisits.reverse();
+
+  // pull out the bits we care about
+  const responseVisitsToCompare: ComparableVisit[] = [];
+  for (var i = 0; i < expectedVisits.length; i++) {
+    const expectedVisit = expectedVisits[i];
+    const completeResponseVisit = increasingDateResponseVisits[i];
+    const simplifiedResponseVisit: ComparableVisit = {};
+
+    if (expectedVisit.tag) {
+      simplifiedResponseVisit.tag = completeResponseVisit.what || "<null>";
+    }
+
+    if (expectedVisit.recordings) {
+      const recordingIds = completeResponseVisit.events.map((ev) => {
+        return ev.recID;
+      });
+      const uniqueRecordingIds = recordingIds.reduce((acc, recId) => {
```

Comment:
This isn't the best way to de-dupe an array.  Better would be to do `Object.keys(recordingIds.reduce((acc, recId) -> { acc[recId] = true; return acc;})) or just use the built-in Set object: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set.

Totally a nit-pick though, and really is only for readability, this would only be inefficient for really big values of n.

Reply:
I tried set but had ts lib set to es5.   All fixed now.   Yay!